core/src/main/scala/shapeless/lazy.scala

```diff
@@ -46,48 +47,187 @@ object Lazy {
   def values[T <: HList](implicit lv: Lazy[Values[T]]): T = lv.value.values
 
   implicit def mkLazy[I]: Lazy[I] = macro LazyMacros.mkLazyImpl[I]
+
+  trait Cached[+T] extends Lazy[T] {
+    override def map[U](f: T => U): Cached[U] = Cached { f(value) }
+    def flatMap[U](f: T => Cached[U]): Cached[U] = Cached { f(value).value }
+  }
+
+  object Cached {
+    implicit def apply[T](t: => T): Cached[T] =
+      new Cached[T] {
+        val value = t
+      }
+
+    def unapply[T](lt: Cached[T]): Option[T] = Some(lt.value)
+
+    implicit def mkCached[I]: Cached[I] = macro LazyMacros.mkCachedLazyImpl[I]
+  }
+
 }
 
 object lazily {
   def apply[T](implicit lv: Lazy[T]): T = lv.value
 }
 
+trait Strict[+T] extends Serializable {
```

Comment:
In my tests, I cut it in the `genericTC` implicit argument for the `TC` of an hlist, `TC[List[CC] :: HNil]` here, so not the "root". As long as the cycle is cut at one point, no matter which one, it should be safe. No need to enforce additional constraints. (Unless I'm missing something here, which I don't think I do.)